File: app/src/main/java/ru/samtakoy/core/business/CardsInteractor.java
package ru.samtakoy.core.business;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;
import ru.samtakoy.core.database.room.entities.CardEntity;
import ru.samtakoy.core.database.room.entities.QPackEntity;
import ru.samtakoy.core.database.room.entities.ThemeEntity;
import ru.samtakoy.core.database.room.entities.other.QPackWithCardIds;

public interface CardsInteractor {

    Completable clearDb();

    Flowable<CardEntity> getCardRx(Long cardId);

    void deleteCardWithRelations(Long cardId);

    Completable setCardNewQuestionTextRx(Long cardId, String text);

    Completable setCardNewAnswerTextRx(Long cardId, String text);

    Flowable<QPackWithCardIds> getQPackWithCardIds(Long qPackId);

    Flowable<QPackEntity> getQPackRx(Long qPackId);

    Completable deleteQPack(Long qPackId);

    Flowable<List<CardEntity>> getQPackCards(Long qPackId);

    Single<ThemeEntity> addNewTheme(Long parentThemeId, String title);

    Completable deleteTheme(Long themeId);

    Single<ThemeEntity> getTheme(Long themeId);

    Flowable<List<ThemeEntity>> getChildThemesRx(Long themeId);

    Flowable<List<QPackEntity>> getChildQPacksRx(Long themeId);

    Flowable<List<QPackEntity>> getAllQPacksByLastViewDateAsc();

    Flowable<List<QPackEntity>> getAllQPacksByCreationDateDesc();

    Completable addFakeCard(Long qPackId);
}

File: app/src/main/java/ru/samtakoy/core/domain/CardsInteractor.kt
package ru.samtakoy.core.domain

import io.reactivex.Flowable
import io.reactivex.Single
import kotlinx.coroutines.flow.Flow
import ru.samtakoy.core.data.local.database.room.entities.CardEntity
import ru.samtakoy.core.data.local.database.room.entities.QPackEntity
import ru.samtakoy.core.data.local.database.room.entities.ThemeEntity
import ru.samtakoy.core.data.local.database.room.entities.other.QPackWithCardIds
import java.util.Date

interface CardsInteractor {
    suspend fun clearDb()

    fun getCardAsFlow(cardId: Long): Flow<CardEntity?>

    fun deleteCardWithRelations(cardId: Long)

    suspend fun setCardNewQuestionText(cardId: Long, text: String)

    suspend fun setCardNewAnswerText(cardId: Long, text: String)

    fun getQPackWithCardIdsAsFlow(qPackId: Long): Flow<QPackWithCardIds>

    fun getQPackRx(qPackId: Long): Flowable<QPackEntity>

    suspend fun getQPack(qPackId: Long): QPackEntity?

    suspend fun deleteQPack(qPackId: Long)

    suspend fun getQPackCards(qPackId: Long): List<CardEntity>

    suspend fun getQPackCardIds(qPackId: Long): List<Long>

    suspend fun updateQPackViewCount(qPackId: Long, currentTime: Date)

    fun addNewTheme(parentThemeId: Long, title: String): Single<ThemeEntity>

    suspend fun deleteTheme(themeId: Long): Boolean

    suspend fun getTheme(themeId: Long): ThemeEntity?

    fun getChildThemesAsFlow(themeId: Long): Flow<List<ThemeEntity>>

    fun getChildQPacksAsFlow(themeId: Long): Flow<List<QPackEntity>>

    fun getAllQPacksByLastViewDateAscAsFlow(
        searchString: String?,
        onlyFavorites: Boolean
    ): Flow<List<QPackEntity>>

    fun getAllQPacksByCreationDateDescAsFlow(
        searchString: String?,
        onlyFavorites: Boolean
    ): Flow<List<QPackEntity>>

    suspend fun getQPacksByIds(ids: List<Long>): List<QPackEntity>

    suspend fun addFakeCard(qPackId: Long)
}
